codice/ddf
# [2.13.X] DDF-3938 Solr query performance improvements

##### ABBREVIATED REVIEW BETWEEN 2.13.X AND MASTER IS IN EFFECT
Link to Master PR: #3553 
____

#### What does this PR do?
Port @pklinef's changes to the DDF release branch.
Makes changes to Solr schema.xml and solrconfig.xml to improve query performance. Enabled docValues (improved sorting and faceting), switched to RptWithGeometrySpatialField (improved Intersects), replaced Trie fields with Point fields (improved range queries), removed sort keys with new multivalued doc value sorting (smaller ingest/index), replaced guava cache with Solr real time get (enable longer auto soft commit), and increased auto commit times (improved Solr cache hit ratio).

#### Who is reviewing it? 
@brendan-hofmann 
@paouelle 

#### Select relevant component teams: 
@codice/solr 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@clockard
@rzwiefel 

#### How should this be tested?
Compare query performance with large dataset before and after. Constantly ingest data (at least one per second or faster) during query performance testing to verify improved caching. I used Apache Bench to get the mean response time and response percentiles but JMeter and others should work too. Here is an example of doing a query for test, sorting by a date, and default page size of 250 as generated by Intrigue. Using 10 concurrent workers over 5000 requests.

```ab -n 5000 -c 10 "http://localhost:8994/solr/catalog/select?q=+(++(*:*+-validation-errors_txt:[*+TO+*])++AND++(++(+metacard-tags_txt_tokenized:%22resource%22+OR++(*:*+-metacard-tags_txt:[*+TO+*])++)++AND+(point-of-contact_txt_ws:%22test%22+source-id_txt_ws:%22test%22+ext.feature-code_txt_ws:%22test%22+contact.publisher-name_txt_ws:%22test%22+validation-errors_txt_ws:%22test%22+media.type_txt_ws:%22test%22+ext.feature-class_txt_ws:%22test%22+id_txt_ws:%22test%22+contact.creator-email_txt_ws:%22test%22+contact.creator-address_txt_ws:%22test%22+checksum_txt_ws:%22test%22+metacard.version.edited-by_txt_ws:%22test%22+checksum-algorithm_txt_ws:%22test%22+metacard.version.resource-uri_txt_ws:%22test%22+metadata-content-type-version_txt_ws:%22test%22+security.access-groups_txt_ws:%22test%22+title_txt_ws:%22test%22+validation-warnings_txt_ws:%22test%22+media.format_txt_ws:%22test%22+metacard.owner_txt_ws:%22test%22+contact.publisher-address_txt_ws:%22test%22+ext.import-location_txt_ws:%22test%22+contact.creator-phone_txt_ws:%22test%22+contact.publisher-email_txt_ws:%22test%22+topic.keyword_txt_ws:%22test%22+security.access-individuals_txt_ws:%22test%22+failed-validators-warnings_txt_ws:%22test%22+topic.category_txt_ws:%22test%22+media.encoding_txt_ws:%22test%22+resource.derived-download-url_txt_ws:%22test%22+metacard.associations.derived_txt_ws:%22test%22+contact.contributor-email_txt_ws:%22test%22+contact.point-of-contact-name_txt_ws:%22test%22+failed-validators-errors_txt_ws:%22test%22+metacard.version.type_txt_ws:%22test%22+resource-size_txt_ws:%22test%22+media.compression_txt_ws:%22test%22+metacard.associations.related_txt_ws:%22test%22+metadata-content-type_txt_ws:%22test%22+location.country-code_txt_ws:%22test%22+contact.creator-name_txt_ws:%22test%22+contact.contributor-phone_txt_ws:%22test%22+contact.publisher-phone_txt_ws:%22test%22+metadata_txt_ws:%22test%22+topic.vocabulary_txt_ws:%22test%22+contact.point-of-contact-email_txt_ws:%22test%22+metacard.version.tags_txt_ws:%22test%22+contact.point-of-contact-address_txt_ws:%22test%22+metacard.version.action_txt_ws:%22test%22+language_txt_ws:%22test%22+metacard.version.id_txt_ws:%22test%22+datatype_txt_ws:%22test%22+media.scanning-mode_txt_ws:%22test%22+description_txt_ws:%22test%22+associations.external_txt_ws:%22test%22+datetime.name_txt_ws:%22test%22+metacard-tags_txt_ws:%22test%22+resource-uri_txt_ws:%22test%22+contact.contributor-name_txt_ws:%22test%22+location.crs-name_txt_ws:%22test%22+contact.point-of-contact-phone_txt_ws:%22test%22+resource.derived-uri_txt_ws:%22test%22+resource-download-url_txt_ws:%22test%22+contact.contributor-address_txt_ws:%22test%22+metadata-target-namespace_txt_ws:%22test%22+media.format-version_txt_ws:%22test%22+location.crs-code_txt_ws:%22test%22)+)++)+&fl=point-of-contact_txt,ext.feature-code_txt,source-id_txt,contact.publisher-name_txt,validation-errors_txt,media.type_txt,ext.feature-class_txt,id_txt,contact.creator-email_txt,metacard_type_name_txt,contact.creator-address_txt,checksum_txt,metacard.version.edited-by_txt,metacard.sharing_xml,checksum-algorithm_txt,metacard.version.resource-uri_txt,queries_xml,security.access-groups_txt,title_txt,validation-warnings_txt,media.format_txt,metacard.owner_txt,contact.publisher-address_txt,ext.import-location_txt,contact.creator-phone_txt,security.access-individuals_txt,topic.keyword_txt,contact.publisher-email_txt,topic.category_txt,media.encoding_txt,failed-validators-warnings_txt,resource.derived-download-url_txt,metacard.associations.derived_txt,contact.contributor-email_txt,contact.point-of-contact-name_txt,metacard.version.type_txt,failed-validators-errors_txt,media.compression_txt,resource-size_txt,metacard.associations.related_txt,location.country-code_txt,contact.creator-name_txt,contact.contributor-phone_txt,contact.publisher-phone_txt,topic.vocabulary_txt,contact.point-of-contact-email_txt,metacard.version.tags_txt,metacard.version.action_txt,contact.point-of-contact-address_txt,language_txt,metacard.version.id_txt,media.scanning-mode_txt,datatype_txt,description_txt,associations.external_txt,datetime.name_txt,metacard-tags_txt,resource-uri_txt,contact.contributor-name_txt,location.crs-name_txt,contact.point-of-contact-phone_txt,resource-download-url_txt,resource.derived-uri_txt,contact.contributor-address_txt,lists_xml,media.format-version_txt,location.crs-code_txt,*_geo,*_bln,*_tdt,*_shr,*_flt,*_dbl,*_int,*_lng,*_bin,*_obj,score&start=0&sort=modified_tdt+desc&rows=250&wt=javabin&version=2"```

#### What are the relevant tickets?
[DDF-3938](https://codice.atlassian.net/browse/DDF-3938)

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
